### PR TITLE
Re-do: Weights & Biases (wandb) integration

### PR DESCRIPTION
**Pull request by alejandromarcu**
This is an improved version of https://github.com/jonbinney/deep_rabbit_hole/pull/144

This PR allows to track the training in wandb, as well as save and load model parameters and configuration.

**Usage**
If you don't want to use wandb during training (e.g. you're just doing some quick tests and care more about the code running than actually storing the results), either pass to the training `--no-wandb` or run from the command line `wand disabled` and then it won't be used until re-enabled

Otherwise, the same train command than before will log to wandb and provide a link to see the run. 

When using `play.py`:
- if you don't provide any additional information to the player (e.g. just `dexp`), then it will try to load the file as if it was saved without wandb (ending in `_final.py`).  If you run the training with wandb, you should use wandb in play as well.
- You can specify a local file to load, e.g. `dexp:model_filename=models/dexp_B7W0_mv1_C110_final.pt`
- In order to load a model from wandb, you need to pass the alias.  Each model can have more than 1 alias (but no alias can be in more than 1 model).  For example, by default you'll have the aliases v0, v1, v2...etc, increasing in each new model, and latest for the last one.  So, if you just want to use the last one, you can pass it as follows: `dexp:wandb_alias=latest`.  You could also choose a specific one, e.g. v0, or manually label something as `best` or whatever.
  - The "play parameters" are stored in wandb metadata when running the training and will be restored when loading.  E.g. if you train a model using `use_rotate_board=True` in DExpAgent, then when the model is loaded from wandb, the agent will have that configuration set.  If loading from a local file, you need to specify the parameters manually. (If useful, we could save the configuration in a local file in the future)
  - Notice that the wandb files are stored locally in the models directory to avoid re-downloading every time, using the first 5 digits of the digest in the name, e.g. `dexp_B5W5_mv1_b0c11.pt`, this way, if you're trying to load `latest` but the alias was moved to a different model, it will recognize that it's different and download it.

### CODE REVIEW

File: deep_quoridor/src/agents/core/trainable_agent.py [hunk 1]
```diff
@@ -293,10 +326,57 @@ def load_model(self, path):
...
+            filename = self.params.model_filename
         else:
-            raise FileNotFoundError(f"Model file {model_path} not found. Please provide the weights file.")
+            # If no filename is passed in training mode, assume we are not loading a model
+            if self.training_mode:
+                return
+
+            # If it's not training mode, we definitely need to load a pretrained model, so try the
+            # default path for local files
+            filename = Path(__file__).resolve().parents[4] / "models" / self.resolve_filename("final")
```
Comment (muralx) [inline deep_quoridor/src/agents/core/trainable_agent.py:340]: maybe make this an class argument too, we should not be guessing where to find the models.  I mean the resolve()..../models/

File: deep_quoridor/src/agents/core/trainable_agent.py [hunk 2]
```diff
@@ -293,10 +326,57 @@ def load_model(self, path):
...
+        The params are updated to the artifact metadata.
+
+        """
+        alias = self.params.wandb_alias
+        if not alias:
+            return
+
+        api = wandb.Api()
+        artifact = api.artifact(f"the-lazy-learning-lair/deep_quoridor/{self.model_id()}:{alias}", type="model")
+        local_filename = Path(__file__).resolve().parents[4] / "models" / self.wandb_local_filename(artifact)
```
Comment (muralx) [inline deep_quoridor/src/agents/core/trainable_agent.py:364]: nit: also make configurable at least the root:
- we can have <path>/models, <path>/wandbmodels (the latter we know they are backed up)
-

File: deep_quoridor/src/agents/core/trainable_agent.py [hunk 3]
```diff
@@ -293,10 +326,57 @@ def load_model(self, path):
...
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            artifact_dir = artifact.download(root=tmpdir)
+
+            path = artifact.download(root=artifact_dir)
+            tmp_filename = Path(path) / f"{self.model_id()}.pt"
+            if not os.path.exists(tmp_filename):
+                raise FileNotFoundError(f"Model file {tmp_filename} was not downloaded.  Please check the artifact")
+
+            os.rename(tmp_filename, local_filename)
```
Comment (muralx) [inline deep_quoridor/src/agents/core/trainable_agent.py:380]: does this work in windows if different disks are used?

File: deep_quoridor/src/plugins/wandb_train.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        if self.episode_count % self.update_every == 0 or self.episode_count == (self.total_episodes - 1):
+            avg_reward, avg_loss = self.agent.compute_loss_and_reward(self.update_every)
+
+            self.run.log(
+                {"loss": avg_loss, "reward": avg_reward, "epsilon": self.agent.epsilon},
+                step=self.episode_count,
+            )
+        self.episode_count += 1
+
+    def end_arena(self, game, results):
```
Comment (muralx) [inline deep_quoridor/src/plugins/wandb_train.py:44]: call wandb.finish()?

File: deep_quoridor/src/train.py [hunk 5]
```diff
@@ -55,16 +56,20 @@ def train_dqn(
...
     agent2.final_reward_multiplier = 2
     # agent2.use_opponentns_actions = False
     # agent2.load_model("models/dexp_B5W0_base.pt")
     agent3 = GreedyAgent()  # noqa: F841
 
-    save_plugin = SaveModelEveryNEpisodesPlugin(
-        update_every=save_frequency, path=save_path, agents=[agent2], board_size=board_size, max_walls=max_walls
-    )
+    if use_wandb:
+        save_plugin = WandbTrainPlugin(update_every=10, path=save_path, agent=agent2, total_episodes=episodes)
```
Comment (muralx) [inline deep_quoridor/src/train.py:67]: can we use the sabemodelevery plugin and only have the Wandb plugin to save the model into wandb after it was saved by the file plugging.

I think we want to keep the save plugin to have checkpoints as we are doing, if we run a 10k training we want to have models saved every x episodes.

**Review by muralx — APPROVED** [review]
Thanks Cucu

File: deep_quoridor/src/agents/core/trainable_agent.py (hunk 1 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/trainable_agent.py]: Done, now you can pass `model_dir` or `wandb_dir` to specify the relative (to the root) or absolute path, and if omitted it will use "models" and "wandbmodels"

File: deep_quoridor/src/agents/core/trainable_agent.py (hunk 3 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/trainable_agent.py:387]: It wouldn't work, but I changed it so that the file is saved in the wandb directory so then the rename is just the name of the file

File: deep_quoridor/src/plugins/wandb_train.py (hunk 4 above)
Comment (alejandromarcu) [reply deep_quoridor/src/plugins/wandb_train.py:44]: Good catch! now wandb is showing some more information when closing
